File: Ch02/Ex2_07-10.scala
/**
  * Ex2_07.scala
  * Kevin C. Baird
  * SICP in Scala (http://www.scala-lang.org/documentation/)
  *
  * SICP in Scala:
  *   Exercise 2.7  - Interval Arithmetic
  *   Exercise 2.8  - Add sub-interval
  *   Exercise 2.9  - Add width
  *   Exercise 2.10 - Add _spans_zero error handling
  */

case class SpansZeroException(message: String) extends Exception(message)
case class Interval(head: Double, end: Double) {
  override def toString(): String = "Interval(" + head + ", " + end + ")";
  def add(otherI: Interval): Interval = {
    val newH = head + otherI.head
    val newE = end  + otherI.end
    Interval(newH, newE)
  }
  def div(otherI: Interval): Interval = {
    val spansZero = 
      (head >= 0 && end <= 0) ||
      (head <= 0 && end >= 0)
    if (spansZero)
      throw SpansZeroException("")
    val recipH = (1.0 / otherI.head): Double
    val recipE = (1.0 / otherI.end): Double
    val recip  = Interval(recipH, recipE)
    mult(recip)
  }
  def mult(otherI: Interval): Interval = {
    val p1 = head * otherI.head
    val p2 = head * otherI.end
    val p3 = end  * otherI.head
    val p4 = end  * otherI.end
    val newH = List(p1, p2, p3, p4).min
    val newE = List(p1, p2, p3, p4).max
    Interval(newH, newE)
  }
  def sub(otherI: Interval): Interval = {
    val newH = head - otherI.head
    val newE = end  - otherI.end
    Interval(newH, newE)
  }
  def width: Double = { (end - head) / 2.0 }
}

def output() = {
  val i1 = new Interval(2, 9)
  val i2 = new Interval(3, 5)
  val i3 = new Interval(-1, 5)
  (
    i1, i2,
    i1.add(i2),
    i1.sub(i2),
    i1.mult(i2),
    i1.div(i2),
    i1.width
    /*
     * error properly raised by i3.div(i1)
     */
  )
}

/**
  * execute with
  * scala -i Ex2_07-10.scala -e "println(output)"
  * until I figure out a better way
  */

File: Ch02/ex2_07-10.rb
#!/usr/bin/env /ruby
# ex2_07-10.rb

=begin rdoc
Kevin C. Baird
SICP in Ruby
Exercise 2.7  - Interval Arithmetic
Exercise 2.8  - Add subtraction
Exercise 2.9  - Add width
Exercise 2.10 - Add spans_zero? error handling
=end

require 'rspec'

class Range

  def +(other_range)
    new_first = (first + other_range.first)
    new_last  = (last  + other_range.last)
    new_first..new_last
  end

  def -(other_range)
    new_first = (first - other_range.first)
    new_last  = (last  - other_range.last)
    new_first..new_last
  end

  def *(other_range)
    p1 = first * other_range.first
    p2 = first * other_range.last
    p3 = last  * other_range.first
    p4 = last  * other_range.last
    new_first = [p1, p2, p3, p4].min
    new_last  = [p1, p2, p3, p4].max
    new_first..new_last
  end

  def /(other_range)
    raise ZeroDivisionError if other_range.spans_zero?
    recip = (1.0 / other_range.last)..(1.0 / other_range.first)
    self * recip
  end

  def width
    (last - first) / 2.0
  end

  protected

  def spans_zero?
    (first >= 0 and last <= 0) or
    (first <= 0 and last >= 0)
  end

end

describe Range do
  r1 = (0..9)
  r2 = (3..5)
  r3 = (0..1)
  describe "#{r1} + #{r2}" do
    subject { r1 + r2 }
    it { should eq((3..14)) }
  end
  describe "#{r1} - #{r2}" do
    subject { r1 - r2 }
    it { should eq((-3..4)) }
  end
  describe "#{r1} * #{r2}" do
    subject { r1 * r2 }
    it { should eq((0..45)) }
  end
  describe "#{r1} / #{r2}" do
    subject { r1 / r2 }
    it { should eq((0..3)) }
  end
  describe "#{r1} / #{r3}" do
    division_by_zero = -> { r1 / r3 }
    it "raises a div by zero error" do
      expect(division_by_zero).to raise_error(ZeroDivisionError)
    end
  end
  describe "#{r1}.width" do
    subject { r1.width }
    it { should eq(4.5) }
  end
  describe "#{r2}.width" do
    subject { r2.width }
    it { should eq(1) }
  end
end

